### PR TITLE
Enhance job cancel commands on a per-user basis

### PR DESCRIPTION
**Pull request by itkovian**
- make sure to add users first
- do not add the cancel commands if there are too many users that would be cancelled
- remove users last (may fail due to cancel commands not being added)

### CODE REVIEW

File: bin/sync_slurm_acct.py [hunk 1]
```diff
@@ -141,21 +141,28 @@ def main():
...
 
+        # Adding users takes priority
+        sacctmgr_commands += user_commands
+
         # safety to avoid emptying the cluster due to some error upstream
-        if not opts.options.force and len(job_cancel_commands > 5):
-            logging.error("Aborting, would otherwise cancel jobs for %d users", len(job_cancel_commands))
-            raise SyncSanityError("Would cancel jobs for %d users" % len(job_cancel_commands))
+        if not opts.options.force and len(job_cancel_commands) > 5:
+            logging.warning("Will not add commands to cancel jobs for %d users", len(job_cancel_commands))
```
Comment (stdweird) [inline bin/sync_slurm_acct.py:157]: fix message to state that this is number of cancel jobs, not user

File: bin/sync_slurm_acct.py [hunk 2]
```diff
@@ -141,21 +141,28 @@ def main():
...
         )
 
+        # Adding users takes priority
+        sacctmgr_commands += user_commands
+
         # safety to avoid emptying the cluster due to some error upstream
-        if not opts.options.force and len(job_cancel_commands > 5):
-            logging.error("Aborting, would otherwise cancel jobs for %d users", len(job_cancel_commands))
-            raise SyncSanityError("Would cancel jobs for %d users" % len(job_cancel_commands))
+        if not opts.options.force and len(job_cancel_commands) > 5:
```
Comment (stdweird) [inline bin/sync_slurm_acct.py:156]: make this at least 30 or so?

File: bin/sync_slurm_acct.py [hunk 3]
```diff
@@ -141,21 +141,28 @@ def main():
...
+            logging.warning("Will not add commands to cancel jobs for %d users", len(job_cancel_commands))
+            logging.debug("Would execute the following cancel commands:")
```
Comment (stdweird) [inline bin/sync_slurm_acct.py:158]: so now, nothing happens at all: no cancelled jobs and no failure reported?

File: bin/sync_slurm_acct.py (hunk 1 above)
Comment (stdweird) [reply bin/sync_slurm_acct.py:157]: how so?

Comment (itkovian) [reply bin/sync_slurm_acct.py:157]: it's a defaultdict, so `len` is the number of keys/values, right?

File: bin/sync_slurm_acct.py [hunk 4]
```diff
@@ -141,28 +143,45 @@ def main():
...
+            execute_commands(sacctmgr_commands)
+
+        # reset to go on with the remainder of the commands
+        sacctmgr_commands = []
+
         # safety to avoid emptying the cluster due to some error upstream
-        if not opts.options.force and len(job_cancel_commands > 5):
-            logging.error("Aborting, would otherwise cancel jobs for %d users", len(job_cancel_commands))
+        if not opts.options.force and len(job_cancel_commands) > MAX_USERS_JOB_CANCEL:
+            logging.warning("Would add commands to cancel jobs for %d users", len(job_cancel_commands))
```
Comment (stdweird) [inline bin/sync_slurm_acct.py:169]: `cancel jobs for %d users` -> `cancel %d jobs for users`

File: bin/sync_slurm_acct.py [hunk 5]
```diff
@@ -141,28 +143,45 @@ def main():
...
+            logging.debug("Would execute the following cancel commands:")
+            for jc in job_cancel_commands.values():
+                logging.debug("%s", jc)
             raise SyncSanityError("Would cancel jobs for %d users" % len(job_cancel_commands))
```
Comment (stdweird) [inline bin/sync_slurm_acct.py:173]: same `cancel jobs for %d users` -> `cancel %d jobs for users`

File: bin/sync_slurm_acct.py (hunk 4 above)
Comment (stdweird) [reply bin/sync_slurm_acct.py:169]: nvm

File: bin/sync_slurm_acct.py [hunk 6]
```diff
@@ -39,6 +39,8 @@
 SYNC_TIMESTAMP_FILENAME = "/var/cache/%s.timestamp" % (NAGIOS_HEADER)
 SYNC_SLURM_ACCT_LOGFILE = "/var/log/%s.log" % (NAGIOS_HEADER)
 
+MAX_USERS_JOB_CANCEL = 30
```
Comment (stdweird) [inline bin/sync_slurm_acct.py:42]: ok, now these are true users; so you can lower it again to 10 or so? not sure what a good limit is here

File: lib/vsc/administration/slurm/sync.py [hunk 7]
```diff
@@ -810,10 +812,13 @@ def slurm_user_accounts(vo_members, active_accounts, slurm_user_info, clusters,
             cluster=cluster,
             default_account=vo_id) for (user, vo_id, _) in new_users
         ])
-        job_cancel_commands.extend([
-            create_remove_user_jobs_command(user=user, cluster=cluster) for user in remove_users
+
+        for user in remove_users:
+            job_cancel_commands[user].append(create_remove_user_jobs_command(user=user, cluster=cluster))
+
+        association_remove_commands.extend([
```
Comment (stdweird) [inline lib/vsc/administration/slurm/sync.py:819]: add a comment `remove user from current VO/...` (if i uunderstand it correclty)

**Comment by stdweird** [comment]
@itkovian minor remarks, this is good to go

File: bin/sync_slurm_acct.py (hunk 6 above)
Comment (itkovian) [reply bin/sync_slurm_acct.py:42]: neither am I :-)

**Pull request merged**